nutils/evaluable.py

```diff
@@ -3532,11 +3549,9 @@ def _node_tuple(self, cache, subgraph, times):
 def _jointdtype(*dtypes):
   'determine joint dtype'
 
-  type_order = bool, int, float, complex
-  kind_order = 'bifc'
-  itype = max(kind_order.index(dtype.kind) if isinstance(dtype,numpy.dtype)
-           else type_order.index(dtype) for dtype in dtypes)
-  return type_order[itype]
+  kind_map = {'b': bool, 'i': int, 'f': float, 'c': complex}
+  kind = max(map(numpy.dtype, dtypes)).kind
```

Comment:
I was not aware that numpy already defines an ordering for dtypes, do you have a reference for that?

Reply:
I don't think there is such an ordering (as `np.dtype` needs to deal with byte order and structured arrays), but this code does work on our restricted set of types: `{bool, int, float, complex}`. I think the 'order' here is roughly `np.can_cast`. If you think it's a bit too magical, we could also use `np.find_common_type([], dtypes)` for a list of types and `np.promote_types(t, s)` for a pair of types.

Reply:
Unless the ordering is a documented feature I think we should not use it. I was not aware of `promote_types` and the other, indeed we could consider switching to that. I'm just not sure if it adds much if we simultaneously want to limit this logic to the four native python types. In fact, thinking about it some more I think we should limit `_joint_dtype` to native type arguments only, as the conversion should already have happened via `function.asdtype`. Any leftover numpy dtype is an error.